tests/test_bounce_integral.py

```diff
@@ -0,0 +1,751 @@
+        np.testing.assert_allclose(bp1, intersect[1::2])
+        np.testing.assert_allclose(bp2, intersect[0::2][1:])
+
+    def test_bp1_before_extrema():
+        start = -np.pi
+        end = -2 * start
+        k = np.linspace(start, end, 5)
+        B = CubicHermiteSpline(
+            k, np.cos(k) + 2 * np.sin(-2 * k), -np.sin(k) - 4 * np.cos(-2 * k)
+        )
+        B_z_ra = B.derivative()
+        pitch = 1 / B(B_z_ra.roots(extrapolate=False))[3]
+        bp1, bp2 = bounce_points(pitch, k, B.c, B_z_ra.c, check=True)
+        bp1, bp2 = map(_filter_not_nan, (bp1, bp2))
+        assert bp1.size and bp2.size
+        # Our routine correctly detects intersection, while scipy, jnp.root fails.
+        intersect = B.solve(1 / pitch, extrapolate=False)
+        np.testing.assert_allclose(bp1[1], 1.9827671337414938)
+        intersect = np.insert(intersect, np.searchsorted(intersect, bp1[1]), bp1[1])
+        np.testing.assert_allclose(bp1, intersect[[1, 2]])
+        np.testing.assert_allclose(bp2, intersect[[2, 3]])
+
+    def test_bp2_before_extrema():
+        start = -1.2 * np.pi
+        end = -2 * start
+        k = np.linspace(start, end, 7)
+        B = CubicHermiteSpline(
+            k,
+            np.cos(k) + 2 * np.sin(-2 * k) + k / 4,
+            -np.sin(k) - 4 * np.cos(-2 * k) + 1 / 4,
+        )
+        B_z_ra = B.derivative()
+        pitch = 1 / B(B_z_ra.roots(extrapolate=False))[2]
+        bp1, bp2 = bounce_points(pitch, k, B.c, B_z_ra.c, check=True)
+        bp1, bp2 = map(_filter_not_nan, (bp1, bp2))
+        assert bp1.size and bp2.size
+        intersect = B.solve(1 / pitch, extrapolate=False)
+        np.testing.assert_allclose(bp1, intersect[[0, -2]])
+        np.testing.assert_allclose(bp2, intersect[[1, -1]])
+
+    def test_extrema_first_and_before_bp1(plot=False):
+        start = -1.2 * np.pi
+        end = -2 * start
+        k = np.linspace(start, end, 7)
+        B = CubicHermiteSpline(
+            k,
+            np.cos(k) + 2 * np.sin(-2 * k) + k / 20,
+            -np.sin(k) - 4 * np.cos(-2 * k) + 1 / 20,
+        )
+        B_z_ra = B.derivative()
+        pitch = 1 / B(B_z_ra.roots(extrapolate=False))[2]
+        bp1, bp2 = bounce_points(pitch, k[2:], B.c[:, 2:], B_z_ra.c[:, 2:], check=True)
+        if plot:
+            plot_field_line_with_ripple(B, pitch, bp1, bp2, start=k[2])
+        bp1, bp2 = map(_filter_not_nan, (bp1, bp2))
+        assert bp1.size and bp2.size
+        # Our routine correctly detects intersection, while scipy, jnp.root fails.
+        intersect = B.solve(1 / pitch, extrapolate=False)
+        np.testing.assert_allclose(bp1[0], 0.8353192766102349)
+        intersect = np.insert(intersect, np.searchsorted(intersect, bp1[0]), bp1[0])
+        intersect = intersect[intersect >= k[2]]
+        np.testing.assert_allclose(bp1, intersect[[0, 1, 3]])
+        np.testing.assert_allclose(bp2, intersect[[0, 2, 4]])
+
+    def test_extrema_first_and_before_bp2():
```

Comment:
Is this the one that tests the first extremum that is paired with the last extremum?

Reply:
One can add the plot=True keyword argument to the `bounce_points` method in each of these tests to generate plots on what the test is doing. 

We are checking that all the correct bounce points are detected for a given pitch.
In this case, the first intersection is an extrema (and so is the second).
![Figure_1](https://github.com/PlasmaControl/DESC/assets/98659578/711107da-1795-4389-9f3a-ca0d20b24e5a)
We need to make sure we are integrating from red to green every time, and never beneath the well from green to red, which the plot confirms.

Note that one could reflect the plot over the purple pitch line. In that case the intersection point is still an extrema, but we would not consider it a bounce point because the line connecting to the next intersection would be beneath the well. These tests ensure such edge cases are correct, and that the code can correctly determine when an extrema should be a bounce point or not.